File: apps/api/internal/model/pain_location.go
package model

import (
	"encoding/json"
)

// PainRegion represents a single pain location marked on the anatomy diagram.
type PainRegion struct {
	ID          string `json:"id" validate:"required"`
	Severity    int    `json:"severity" validate:"min=0,max=10"`
	Description string `json:"description,omitempty" validate:"max=500"`
}

// PainLocationData holds all pain locations for a visit.
// Stored as JSONB in the treatment_sessions table.
type PainLocationData struct {
	Regions []PainRegion `json:"regions"`
}

// PainLocationDataFromJSON parses JSONB data into PainLocationData.
func PainLocationDataFromJSON(data json.RawMessage) (*PainLocationData, error) {
	if data == nil || len(data) == 0 {
		return &PainLocationData{Regions: []PainRegion{}}, nil
	}
	var pld PainLocationData
	if err := json.Unmarshal(data, &pld); err != nil {
		return nil, err
	}
	if pld.Regions == nil {
		pld.Regions = []PainRegion{}
	}
	return &pld, nil
}

// ToJSON serializes PainLocationData to JSON for database storage.
func (p *PainLocationData) ToJSON() (json.RawMessage, error) {
	if p == nil {
		return json.Marshal(&PainLocationData{Regions: []PainRegion{}})
	}
	if p.Regions == nil {
		p.Regions = []PainRegion{}
	}
	return json.Marshal(p)
}

// UpdatePainLocationsRequest is the request body for updating pain locations on a visit.
type UpdatePainLocationsRequest struct {
	SessionID string       `json:"session_id" validate:"required,uuid"`
	Regions   []PainRegion `json:"regions" validate:"required,dive"`
}

// PainLocationsResponse is the API response for pain location data.
type PainLocationsResponse struct {
	SessionID string       `json:"session_id"`
	Regions   []PainRegion `json:"regions"`
}

// ValidAnatomyRegions lists all valid anatomy region IDs.
var ValidAnatomyRegions = map[string]bool{
	// Front view
	"head":                    true,
	"neck_front":              true,
	"shoulder_left":           true,
	"shoulder_right":          true,
	"upper_arm_left_front":    true,
	"upper_arm_right_front":   true,
	"elbow_left":              true,
	"elbow_right":             true,
	"forearm_left":            true,
	"forearm_right":           true,
	"wrist_hand_left":         true,
	"wrist_hand_right":        true,
	"chest_left":              true,
	"chest_right":             true,
	"abdomen_upper":           true,
	"abdomen_lower":           true,
	"hip_left":                true,
	"hip_right":               true,
	"groin_left":              true,
	"groin_right":             true,
	"thigh_left_front":        true,
	"thigh_right_front":       true,
	"knee_left":               true,
	"knee_right":              true,
	"lower_leg_left_front":    true,
	"lower_leg_right_front":   true,
	"ankle_left":              true,
	"ankle_right":             true,
	"foot_left":               true,
	"foot_right":              true,
	// Back view
	"neck_back":               true,
	"cervical_spine":          true,
	"thoracic_spine_upper":    true,
	"thoracic_spine_lower":    true,
	"lumbar_spine":            true,
	"sacrum":                  true,
	"upper_arm_left_back":     true,
	"upper_arm_right_back":    true,
	"gluteal_left":            true,
	"gluteal_right":           true,
	"thigh_left_back":         true,
	"thigh_right_back":        true,
	"knee_left_back":          true,
	"knee_right_back":         true,
	"calf_left":               true,
	"calf_right":              true,
}

// ValidateRegionID checks if a region ID is valid.
func ValidateRegionID(id string) bool {
	return ValidAnatomyRegions[id]
}

// AnatomyRegion represents detailed metadata for an anatomy region.
type AnatomyRegion struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	NameVi      string `json:"name_vi"`
	Category    string `json:"category"` // "upper_limb", "lower_limb", "spine", "trunk", "head_neck"
	View        string `json:"view"`     // "front" or "back"
	Side        string `json:"side"`     // "left", "right", or "bilateral"
	Description string `json:"description,omitempty"`
}

// AnatomyRegionsMetadata contains detailed information for all anatomy regions.
var AnatomyRegionsMetadata = []AnatomyRegion{
	// Head and Neck - Front View
	{ID: "head", Name: "Head", NameVi: "Đầu", Category: "head_neck", View: "front", Side: "bilateral", Description: "Cranial region including forehead, temples, and scalp"},
	{ID: "neck_front", Name: "Neck (Front)", NameVi: "Cổ (Phía trước)", Category: "head_neck", View: "front", Side: "bilateral", Description: "Anterior cervical region"},

	// Upper Limb - Front View
	{ID: "shoulder_left", Name: "Left Shoulder", NameVi: "Vai trái", Category: "upper_limb", View: "front", Side: "left", Description: "Left glenohumeral joint and deltoid region"},
	{ID: "shoulder_right", Name: "Right Shoulder", NameVi: "Vai phải", Category: "upper_limb", View: "front", Side: "right", Description: "Right glenohumeral joint and deltoid region"},
	{ID: "upper_arm_left_front", Name: "Left Upper Arm (Front)", NameVi: "Cánh tay trên trái (Phía trước)", Category: "upper_limb", View: "front", Side: "left", Description: "Left anterior brachial region"},
	{ID: "upper_arm_right_front", Name: "Right Upper Arm (Front)", NameVi: "Cánh tay trên phải (Phía trước)", Category: "upper_limb", View: "front", Side: "right", Description: "Right anterior brachial region"},
	{ID: "elbow_left", Name: "Left Elbow", NameVi: "Khuỷu tay trái", Category: "upper_limb", View: "front", Side: "left", Description: "Left cubital region"},
	{ID: "elbow_right", Name: "Right Elbow", NameVi: "Khuỷu tay phải", Category: "upper_limb", View: "front", Side: "right", Description: "Right cubital region"},
	{ID: "forearm_left", Name: "Left Forearm", NameVi: "Cẳng tay trái", Category: "upper_limb", View: "front", Side: "left", Description: "Left antebrachial region"},
	{ID: "forearm_right", Name: "Right Forearm", NameVi: "Cẳng tay phải", Category: "upper_limb", View: "front", Side: "right", Description: "Right antebrachial region"},
	{ID: "wrist_hand_left", Name: "Left Wrist & Hand", NameVi: "Cổ tay & Bàn tay trái", Category: "upper_limb", View: "front", Side: "left", Description: "Left carpal and hand region"},
	{ID: "wrist_hand_right", Name: "Right Wrist & Hand", NameVi: "Cổ tay & Bàn tay phải", Category: "upper_limb", View: "front", Side: "right", Description: "Right carpal and hand region"},

	// Trunk - Front View
	{ID: "chest_left", Name: "Left Chest", NameVi: "Ngực trái", Category: "trunk", View: "front", Side: "left", Description: "Left thoracic region"},
	{ID: "chest_right", Name: "Right Chest", NameVi: "Ngực phải", Category: "trunk", View: "front", Side: "right", Description: "Right thoracic region"},
	{ID: "abdomen_upper", Name: "Upper Abdomen", NameVi: "Bụng trên", Category: "trunk", View: "front", Side: "bilateral", Description: "Epigastric and umbilical region"},
	{ID: "abdomen_lower", Name: "Lower Abdomen", NameVi: "Bụng dưới", Category: "trunk", View: "front", Side: "bilateral", Description: "Hypogastric and pelvic region"},

	// Lower Limb - Front View
	{ID: "hip_left", Name: "Left Hip", NameVi: "Hông trái", Category: "lower_limb", View: "front", Side: "left", Description: "Left hip joint and anterior pelvic region"},
	{ID: "hip_right", Name: "Right Hip", NameVi: "Hông phải", Category: "lower_limb", View: "front", Side: "right", Description: "Right hip joint and anterior pelvic region"},
	{ID: "groin_left", Name: "Left Groin", NameVi: "Bẹn trái", Category: "lower_limb", View: "front", Side: "left", Description: "Left inguinal region"},
	{ID: "groin_right", Name: "Right Groin", NameVi: "Bẹn phải", Category: "lower_limb", View: "front", Side: "right", Description: "Right inguinal region"},
	{ID: "thigh_left_front", Name: "Left Thigh (Front)", NameVi: "Đùi trái (Phía trước)", Category: "lower_limb", View: "front", Side: "left", Description: "Left anterior femoral region"},
	{ID: "thigh_right_front", Name: "Right Thigh (Front)", NameVi: "Đùi phải (Phía trước)", Category: "lower_limb", View: "front", Side: "right", Description: "Right anterior femoral region"},
	{ID: "knee_left", Name: "Left Knee", NameVi: "Gối trái", Category: "lower_limb", View: "front", Side: "left", Description: "Left patellofemoral and tibiofemoral joint"},
	{ID: "knee_right", Name: "Right Knee", NameVi: "Gối phải", Category: "lower_limb", View: "front", Side: "right", Description: "Right patellofemoral and tibiofemoral joint"},
	{ID: "lower_leg_left_front", Name: "Left Lower Leg (Front)", NameVi: "Chân dưới trái (Phía trước)", Category: "lower_limb", View: "front", Side: "left", Description: "Left anterior tibial region"},
	{ID: "lower_leg_right_front", Name: "Right Lower Leg (Front)", NameVi: "Chân dưới phải (Phía trước)", Category: "lower_limb", View: "front", Side: "right", Description: "Right anterior tibial region"},
	{ID: "ankle_left", Name: "Left Ankle", NameVi: "Cổ chân trái", Category: "lower_limb", View: "front", Side: "left", Description: "Left ankle joint"},
	{ID: "ankle_right", Name: "Right Ankle", NameVi: "Cổ chân phải", Category: "lower_limb", View: "front", Side: "right", Description: "Right ankle joint"},
	{ID: "foot_left", Name: "Left Foot", NameVi: "Bàn chân trái", Category: "lower_limb", View: "front", Side: "left", Description: "Left foot"},
	{ID: "foot_right", Name: "Right Foot", NameVi: "Bàn chân phải", Category: "lower_limb", View: "front", Side: "right", Description: "Right foot"},

	// Head and Neck - Back View
	{ID: "neck_back", Name: "Neck (Back)", NameVi: "Cổ (Phía sau)", Category: "head_neck", View: "back", Side: "bilateral", Description: "Posterior cervical region"},

	// Spine - Back View
	{ID: "cervical_spine", Name: "Cervical Spine", NameVi: "Cột sống cổ", Category: "spine", View: "back", Side: "bilateral", Description: "C1-C7 vertebrae"},
	{ID: "thoracic_spine_upper", Name: "Upper Thoracic Spine", NameVi: "Cột sống ngực trên", Category: "spine", View: "back", Side: "bilateral", Description: "T1-T6 vertebrae"},
	{ID: "thoracic_spine_lower", Name: "Lower Thoracic Spine", NameVi: "Cột sống ngực dưới", Category: "spine", View: "back", Side: "bilateral", Description: "T7-T12 vertebrae"},
	{ID: "lumbar_spine", Name: "Lumbar Spine", NameVi: "Cột sống thắt lưng", Category: "spine", View: "back", Side: "bilateral", Description: "L1-L5 vertebrae"},
	{ID: "sacrum", Name: "Sacrum", NameVi: "Xương cùng", Category: "spine", View: "back", Side: "bilateral", Description: "Sacral region"},

	// Upper Limb - Back View
	{ID: "upper_arm_left_back", Name: "Left Upper Arm (Back)", NameVi: "Cánh tay trên trái (Phía sau)", Category: "upper_limb", View: "back", Side: "left", Description: "Left posterior brachial region"},
	{ID: "upper_arm_right_back", Name: "Right Upper Arm (Back)", NameVi: "Cánh tay trên phải (Phía sau)", Category: "upper_limb", View: "back", Side: "right", Description: "Right posterior brachial region"},

	// Lower Limb - Back View
	{ID: "gluteal_left", Name: "Left Gluteal", NameVi: "Mông trái", Category: "lower_limb", View: "back", Side: "left", Description: "Left buttock region"},
	{ID: "gluteal_right", Name: "Right Gluteal", NameVi: "Mông phải", Category: "lower_limb", View: "back", Side: "right", Description: "Right buttock region"},
	{ID: "thigh_left_back", Name: "Left Thigh (Back)", NameVi: "Đùi trái (Phía sau)", Category: "lower_limb", View: "back", Side: "left", Description: "Left posterior femoral region"},
	{ID: "thigh_right_back", Name: "Right Thigh (Back)", NameVi: "Đùi phải (Phía sau)", Category: "lower_limb", View: "back", Side: "right", Description: "Right posterior femoral region"},
	{ID: "knee_left_back", Name: "Left Knee (Back)", NameVi: "Gối trái (Phía sau)", Category: "lower_limb", View: "back", Side: "left", Description: "Left popliteal region"},
	{ID: "knee_right_back", Name: "Right Knee (Back)", NameVi: "Gối phải (Phía sau)", Category: "lower_limb", View: "back", Side: "right", Description: "Right popliteal region"},
	{ID: "calf_left", Name: "Left Calf", NameVi: "Bắp chân trái", Category: "lower_limb", View: "back", Side: "left", Description: "Left posterior leg region"},
	{ID: "calf_right", Name: "Right Calf", NameVi: "Bắp chân phải", Category: "lower_limb", View: "back", Side: "right", Description: "Right posterior leg region"},
}

// GetAllAnatomyRegions returns all anatomy regions with metadata.
func GetAllAnatomyRegions() []AnatomyRegion {
	return AnatomyRegionsMetadata
}

// GetAnatomyRegion returns a specific anatomy region by ID.
func GetAnatomyRegion(id string) (AnatomyRegion, bool) {
	for _, region := range AnatomyRegionsMetadata {
		if region.ID == id {
			return region, true
		}
	}
	return AnatomyRegion{}, false
}

File: packages/shared-types/src/pain-location.ts
/**
 * Pain Location types for anatomy visualization
 *
 * Used by the interactive body diagram to mark pain locations
 * with severity ratings and descriptions.
 */

// -----------------------------------------------------------------------------
// Anatomy Region IDs
// -----------------------------------------------------------------------------

/**
 * All clickable anatomy regions available on the body diagrams.
 * Front view regions include anterior body parts.
 * Back view regions include posterior body parts.
 */
export type AnatomyRegionId =
  // Head & Neck
  | 'head'
  | 'neck_front'
  | 'neck_back'
  // Shoulders
  | 'shoulder_left'
  | 'shoulder_right'
  // Upper Arms
  | 'upper_arm_left_front'
  | 'upper_arm_right_front'
  | 'upper_arm_left_back'
  | 'upper_arm_right_back'
  // Elbows
  | 'elbow_left'
  | 'elbow_right'
  // Forearms
  | 'forearm_left'
  | 'forearm_right'
  // Wrists & Hands
  | 'wrist_hand_left'
  | 'wrist_hand_right'
  // Chest & Trunk
  | 'chest_left'
  | 'chest_right'
  | 'abdomen_upper'
  | 'abdomen_lower'
  // Spine
  | 'cervical_spine'
  | 'thoracic_spine_upper'
  | 'thoracic_spine_lower'
  | 'lumbar_spine'
  | 'sacrum'
  // Hips & Pelvis
  | 'hip_left'
  | 'hip_right'
  | 'groin_left'
  | 'groin_right'
  | 'gluteal_left'
  | 'gluteal_right'
  // Thighs
  | 'thigh_left_front'
  | 'thigh_right_front'
  | 'thigh_left_back'
  | 'thigh_right_back'
  // Knees
  | 'knee_left'
  | 'knee_right'
  | 'knee_left_back'
  | 'knee_right_back'
  // Lower Legs
  | 'lower_leg_left_front'
  | 'lower_leg_right_front'
  | 'calf_left'
  | 'calf_right'
  // Ankles & Feet
  | 'ankle_left'
  | 'ankle_right'
  | 'foot_left'
  | 'foot_right';

// -----------------------------------------------------------------------------
// Pain Location Types
// -----------------------------------------------------------------------------

/**
 * A single pain location marked on the anatomy diagram.
 * Named AnatomyPainLocation to avoid conflict with the clinical PainLocation
 * in clinical.ts which has a different, more detailed structure.
 */
export interface AnatomyPainLocation {
  /** Anatomy region identifier */
  id: AnatomyRegionId;
  /** Pain severity 0-10 (VAS scale) */
  severity: number;
  /** Optional description of the pain */
  description?: string;
}

/**
 * Collection of pain locations for a visit.
 * Stored as JSONB in the database.
 */
export interface PainLocationData {
  regions: AnatomyPainLocation[];
}

// -----------------------------------------------------------------------------
// Diagram View Types
// -----------------------------------------------------------------------------

/** Which side of the body diagram to display */
export type AnatomyView = 'front' | 'back';

/**
 * Metadata for an anatomy region used for rendering.
 */
export interface AnatomyRegionMeta {
  id: AnatomyRegionId;
  /** English label */
  label: string;
  /** Vietnamese label */
  labelVi: string;
  /** Which diagram view this region appears on */
  view: AnatomyView;
}

/**
 * Complete registry of all anatomy regions with metadata.
 */
export const ANATOMY_REGIONS: AnatomyRegionMeta[] = [
  // Front view regions
  { id: 'head', label: 'Head', labelVi: 'Dau', view: 'front' },
  { id: 'neck_front', label: 'Neck (Front)', labelVi: 'Co (truoc)', view: 'front' },
  { id: 'shoulder_left', label: 'Left Shoulder', labelVi: 'Vai trai', view: 'front' },
  { id: 'shoulder_right', label: 'Right Shoulder', labelVi: 'Vai phai', view: 'front' },
  { id: 'upper_arm_left_front', label: 'Left Upper Arm', labelVi: 'Canh tay trai', view: 'front' },
  { id: 'upper_arm_right_front', label: 'Right Upper Arm', labelVi: 'Canh tay phai', view: 'front' },
  { id: 'elbow_left', label: 'Left Elbow', labelVi: 'Khuyu tay trai', view: 'front' },
  { id: 'elbow_right', label: 'Right Elbow', labelVi: 'Khuyu tay phai', view: 'front' },
  { id: 'forearm_left', label: 'Left Forearm', labelVi: 'Cang tay trai', view: 'front' },
  { id: 'forearm_right', label: 'Right Forearm', labelVi: 'Cang tay phai', view: 'front' },
  { id: 'wrist_hand_left', label: 'Left Wrist/Hand', labelVi: 'Co tay/Ban tay trai', view: 'front' },
  { id: 'wrist_hand_right', label: 'Right Wrist/Hand', labelVi: 'Co tay/Ban tay phai', view: 'front' },
  { id: 'chest_left', label: 'Left Chest', labelVi: 'Nguc trai', view: 'front' },
  { id: 'chest_right', label: 'Right Chest', labelVi: 'Nguc phai', view: 'front' },
  { id: 'abdomen_upper', label: 'Upper Abdomen', labelVi: 'Bung tren', view: 'front' },
  { id: 'abdomen_lower', label: 'Lower Abdomen', labelVi: 'Bung duoi', view: 'front' },
  { id: 'hip_left', label: 'Left Hip', labelVi: 'Hong trai', view: 'front' },
  { id: 'hip_right', label: 'Right Hip', labelVi: 'Hong phai', view: 'front' },
  { id: 'groin_left', label: 'Left Groin', labelVi: 'Ben trai', view: 'front' },
  { id: 'groin_right', label: 'Right Groin', labelVi: 'Ben phai', view: 'front' },
  { id: 'thigh_left_front', label: 'Left Thigh (Front)', labelVi: 'Dui trai (truoc)', view: 'front' },
  { id: 'thigh_right_front', label: 'Right Thigh (Front)', labelVi: 'Dui phai (truoc)', view: 'front' },
  { id: 'knee_left', label: 'Left Knee', labelVi: 'Goi trai', view: 'front' },
  { id: 'knee_right', label: 'Right Knee', labelVi: 'Goi phai', view: 'front' },
  { id: 'lower_leg_left_front', label: 'Left Shin', labelVi: 'Cang chan trai', view: 'front' },
  { id: 'lower_leg_right_front', label: 'Right Shin', labelVi: 'Cang chan phai', view: 'front' },
  { id: 'ankle_left', label: 'Left Ankle', labelVi: 'Co chan trai', view: 'front' },
  { id: 'ankle_right', label: 'Right Ankle', labelVi: 'Co chan phai', view: 'front' },
  { id: 'foot_left', label: 'Left Foot', labelVi: 'Ban chan trai', view: 'front' },
  { id: 'foot_right', label: 'Right Foot', labelVi: 'Ban chan phai', view: 'front' },
  // Back view regions
  { id: 'neck_back', label: 'Neck (Back)', labelVi: 'Co (sau)', view: 'back' },
  { id: 'cervical_spine', label: 'Cervical Spine', labelVi: 'Cot song co', view: 'back' },
  { id: 'thoracic_spine_upper', label: 'Upper Back', labelVi: 'Lung tren', view: 'back' },
  { id: 'thoracic_spine_lower', label: 'Mid Back', labelVi: 'Lung giua', view: 'back' },
  { id: 'lumbar_spine', label: 'Lower Back', labelVi: 'That lung', view: 'back' },
  { id: 'sacrum', label: 'Sacrum', labelVi: 'Xuong cung', view: 'back' },
  { id: 'upper_arm_left_back', label: 'Left Tricep', labelVi: 'Co tam dau trai', view: 'back' },
  { id: 'upper_arm_right_back', label: 'Right Tricep', labelVi: 'Co tam dau phai', view: 'back' },
  { id: 'gluteal_left', label: 'Left Gluteal', labelVi: 'Mong trai', view: 'back' },
  { id: 'gluteal_right', label: 'Right Gluteal', labelVi: 'Mong phai', view: 'back' },
  { id: 'thigh_left_back', label: 'Left Hamstring', labelVi: 'Dui trai (sau)', view: 'back' },
  { id: 'thigh_right_back', label: 'Right Hamstring', labelVi: 'Dui phai (sau)', view: 'back' },
  { id: 'knee_left_back', label: 'Left Knee (Back)', labelVi: 'Goi trai (sau)', view: 'back' },
  { id: 'knee_right_back', label: 'Right Knee (Back)', labelVi: 'Goi phai (sau)', view: 'back' },
  { id: 'calf_left', label: 'Left Calf', labelVi: 'Bap chan trai', view: 'back' },
  { id: 'calf_right', label: 'Right Calf', labelVi: 'Bap chan phai', view: 'back' },
];

/**
 * Get severity color for rendering on the diagram.
 * Returns a CSS color string based on severity level.
 */
export function getSeverityColor(severity: number): string {
  if (severity <= 0) return 'transparent';
  if (severity <= 2) return 'rgba(255, 235, 59, 0.6)';  // Yellow - mild
  if (severity <= 4) return 'rgba(255, 152, 0, 0.6)';   // Orange - moderate
  if (severity <= 6) return 'rgba(255, 87, 34, 0.6)';   // Deep orange
  if (severity <= 8) return 'rgba(244, 67, 54, 0.6)';   // Red - severe
  return 'rgba(183, 28, 28, 0.7)';                       // Dark red - extreme
}

/**
 * Get severity label for display.
 */
export function getSeverityLabel(severity: number): string {
  if (severity <= 0) return 'None';
  if (severity <= 2) return 'Mild';
  if (severity <= 4) return 'Moderate';
  if (severity <= 6) return 'Moderate-Severe';
  if (severity <= 8) return 'Severe';
  return 'Extreme';
}

/**
 * Get Vietnamese severity label for display.
 */
export function getSeverityLabelVi(severity: number): string {
  if (severity <= 0) return 'Khong dau';
  if (severity <= 2) return 'Nhe';
  if (severity <= 4) return 'Trung binh';
  if (severity <= 6) return 'Trung binh - Nang';
  if (severity <= 8) return 'Nang';
  return 'Rat nang';
}
